caffe2/operators/onnx_while_op.h

```diff
@@ -0,0 +1,224 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CAFFE2_OPERATORS_ONNX_WHILE_OP_H_
+#define CAFFE2_OPERATORS_ONNX_WHILE_OP_H_
+
+#include "caffe2/core/context.h"
+#include "caffe2/core/logging.h"
+#include "caffe2/core/operator.h"
+
+namespace caffe2 {
+
+template <class Context> class ONNXWhileOp final : public Operator<Context> {
+public:
+  ONNXWhileOp(const OperatorDef &operator_def, Workspace *ws)
+      : Operator<Context>(operator_def, ws), ws_(ws),
+        has_trip_count_(
+            OperatorBase::GetSingleArgument<int64_t>("has_trip_count", 0)),
+        has_cond_(OperatorBase::GetSingleArgument<int64_t>("has_cond", 0)) {
+    CAFFE_ENFORCE(this->template HasSingleArgumentOfType<NetDef>("body"),
+                  "body net must be specified in ONNXWhile operator");
+    body_net_def_ = this->template GetSingleArgument<NetDef>("body", NetDef());
+
+    // Create loop-carried deps in Workspace
+    for (int i = 2; i < body_net_def_.external_input_size(); ++i) {
+      Blob *b = ws_->CreateBlob(body_net_def_.external_input()[i]);
+      Tensor<Context> *t = b->template GetMutable<Tensor<Context>>();
+      lcd_tensors_.push_back(t);
+    }
+    // First output is the iteration variable
+    auto *iteration_var_blob = ws_->CreateBlob(body_net_def_.external_input(0));
+    iteration_var_ = iteration_var_blob->template GetMutable<Tensor<Context>>();
+
+    input_condition_var_ = ws_->CreateBlob(body_net_def_.external_input(1))
+                               ->template GetMutable<Tensor<Context>>();
+
+    auto *condition_var_blob =
+        ws_->CreateBlob(body_net_def_.external_output()[0]);
+    condition_var_ = condition_var_blob->template GetMutable<Tensor<Context>>();
+
+    body_net_ = CreateNet(body_net_def_, ws);
+    CAFFE_ENFORCE(body_net_, "Failed to initialize loop subnet");
+  }
+
+  USE_OPERATOR_CONTEXT_FUNCTIONS;
+
+  // Operator
+  //  Inputs: max trip count, condition, initial loop-carried dependencies
+  //  Outputs: Iteration number, Final loop-carried dependencies, scan_outputs
+  // Body
+  //  Inputs: iteration number, condition, loop-carried dependencies
+  //  Outputs: condition, loop-carried dependencies, scan_outputs
+  bool RunOnDevice() override {
+    constexpr int64_t num_inputs_before_lcds = 2;
+    // First input is the maximumt trip count. Second input is the condition
+    // variable (for the first iteration). The rest of the inputs are
+    // loop-carried dependencies.
+    int num_loop_carried_deps = InputSize() - num_inputs_before_lcds;
+    int64_t max_trip_count = *Input(0).template data<int64_t>();
+    const int64_t first_iter_condition = *Input(1).template data<int64_t>();
```

Comment:
I assume there's a conscious decision not to use bool?

Reply:
There was some compilation issue i ran into when i put bool here, let me try to repro